Lms/admin snapshots filters teachers without classrooms

anathomical opened the pull request:
## WHAT
- Update admin filter generation logic to include Teachers that do not have classrooms (as long as they're associated with a relevant school)
- Update the logic for applying filters to API calls so that it better understands when "all" items from a filter are selected because we want to submit `null` as the value for that category filter in those cases instead of having a huge array of redundant IDs
- Update `custom` timeframe calculation so that it has the same start and end of day as other time-frames
## WHY
- We want to count and include teachers in our stats even if they don't have classrooms
- Simplifying the "all selected" logic makes our API calls and caching less complex, and also ensures that we don't mistakenly exclude items in cases where the logic that populates the filters is off but the user doesn't care to apply filters of that class
- We want someone who puts in a custom timeframe to get the same results as a matching pre-defined timeframe
## HOW
- Use `LEFT OUTER JOINS` for `classroom` relationships when generating teacher filters so that if the teacher has no classroom they still end up in the results
- Add an extra condition when deciding whether or not to send `null` as a filter value so that if `selected` items is a superset of `all` items (containing at least every member of `all`), we count that as all-selected even if `selected` has additional items in it.
- Just add `beginning_of_day` and `end_of_day` to custom timeframe calculation lambdas

### Notion Card Links
https://www.notion.so/quill/Admin-Usage-Snapshot-Report-QA-Week-of-July-17-767225ae4fd74382bfed44861d36f589?d=1c41c3ee821141fcacc2c8054d38b703#2fa451e21bca4905bac0a92959ac9097

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? |  Yes
Have you deployed to Staging? | Yes
Self-Review: Have you done an initial self-review of the code below on Github? | Yes
Spec Review: Have you reviewed the spec and ensured this meets requirements and/or matches design mockups? | N/A

anathomical committed:
Include teachers without classrooms in filters

anathomical committed:
Update logic for when "all" items are selected from a filter

Specifically, we allow "selected" items to be a super-set of "all" items in cases where selecting (say) a single school reduces the number of teachers being shown as "all" because we only want to show the teachers from that school.  This code recognizes that "all" are selected if every item in the "all" is included in "selected" even if "selected" has additional items

anathomical committed:
Fix linting error

emilia-friedberg commented on services/QuillLMS/client/app/bundles/PremiumHub/containers/UsageSnapshotsContainer.tsx:
```diff
@@ -239,7 +240,12 @@ const UsageSnapshotsContainer = ({ adminInfo, accessType, }) => {
   function mapItemsIfNotAll(selectedItems, allItems, mapKey = 'id') {
-    if (unorderedArraysAreEqual(selectedItems, allItems)) return null
+    // selectedItems may, by design, be a superset of allItems, but if everything in allItems is in selectedItems, we want to treat it as "everything" being selected
+    const allItemsSelected = allItems.every((i) => {
```
nit, but I always read `i` as `index`.

anathomical replied:
Good call-out.  Worth making that clearer.

anathomical committed:
Tweak timeframe calculation so that custom values go to day boundary

We were just taking raw values from the datepicker, but in order to be consistent with other values, we want to set bounds to start and end of day.  This means that if you do a custom timeframe that corresponds (date-wise) with with a defined timeframe, you should get the same results

anathomical committed:
Name variable something more useful than `i`

happythenewsad commented on services/QuillLMS/client/app/bundles/PremiumHub/containers/UsageSnapshotsContainer.tsx:
```diff
@@ -239,7 +240,12 @@ const UsageSnapshotsContainer = ({ adminInfo, accessType, }) => {
   function mapItemsIfNotAll(selectedItems, allItems, mapKey = 'id') {
-    if (unorderedArraysAreEqual(selectedItems, allItems)) return null
+    // selectedItems may, by design, be a superset of allItems, but if everything in allItems is in selectedItems, we want to treat it as "everything" being selected
+    const allItemsSelected = allItems.every((item) => {
```
Personally I think
`_.intersection(...).length` is more readable and probably a lower runtime complexity, although I didn't check

anathomical replied:
The idea here would be to do something like:
```javascript
const allItemsSelected = _.intersection(selectedItems, allItems).length === allItems.length
```
